pkg/raft/node.go

```diff
@@ -138,6 +125,57 @@ func (m *StorageAppend) Describe(f EntryFormatter) string {
 	return buf.String()
 }
 
+// StorageAppendAck acknowledges that the corresponding StorageAppend is durable
+// on the log storage.
+type StorageAppendAck struct {
+	// Mark is the durable log mark. By the time this acknowledgement is handled,
+	// the log storage can already be at a higher mark.
+	Mark LogMark
+	// SnapIndex is the index of the snapshot that has been applied. If there was
+	// no snapshot, SnapIndex == 0.
+	SnapIndex uint64
+	// responses contains messages that should be sent now that the StorageAppend
+	// is durable. Messages directed to the local RawNode are stepped locally.
+	responses []pb.Message
+}
+
+// Send iterates through the messages that should be sent to remote peers, i.e.
+// peers with ID != self.
+//
+// TODO(pav-kv): in a typical case, all the Responses are addressed to the
+// proposer of the current Term (candidate or leader), and there is no point in
+// sending responses to stale proposers. We can double-down on this, and make an
+// invariant that all the Responses are addressed to the same proposer. Then it
+// is either the local RawNode, or a remote one. So we can avoid scanning the
+// Responses twice (in Send and step).
+func (m *StorageAppendAck) Send(self pb.PeerID) iter.Seq[pb.Message] {
+	return func(yield func(pb.Message) bool) {
+		for _, msg := range m.responses {
+			if msg.To == self {
+				continue
+			} else if !yield(msg) {
+				return
+			}
+		}
+	}
+}
+
+// step iterates through the messages that should be stepped to the local
+// RawNode when applying this acknowledgement.
+func (m *StorageAppendAck) step(self pb.PeerID) iter.Seq[pb.Message] {
```

Comment:
does doing this via in `iter.Seq` allocate (over just iterating over the slice directly)?

Reply:
I did some benchmarking of similar `iter.Seq` things, and it didn't allocate. Go compiler makes good effort to optimize this, last time I read about it. I'll double-check here.

Reply:
```
BenchmarkRawNodeAckAppend   43279309   23.25 ns/op   0 B/op   0 allocs/op
```

```golang
func BenchmarkRawNodeAckAppend(b *testing.B) {
	s := newTestMemoryStorage(withPeers(1))
	rn := newTestRawNode(1, 3, 1, s)

	require.NoError(b, rn.Campaign())
	rd := rn.Ready()
	rn.AckAppend(rd.Ack())
	require.Equal(b, pb.StateLeader, rn.State())

	require.NoError(b, rn.Propose([]byte("proposal")))
	rd = rn.Ready()

	b.ReportAllocs()
	for range b.N {
		rn.AckAppend(rd.Ack())
		rd = rn.Ready()
	}
}
```

Reply:
This might be inaccurate, because `rd` becomes empty after first iteration. Seeing if I can do a better test.

Reply:
TIL, thanks for checking! Do we have any RawNode-specific benchmark? Just curious.

Reply:
Yeah, there is `BenchmarkRawNode`. Don't know if it's functional though. I tried running it on `master` and it doesn't terminate. Will look into it separately.

Reply:
https://github.com/cockroachdb/cockroach/pull/143705#issuecomment-2780146134